File: src/main/kotlin/nl/pindab0ter/common/helpers/adventOfCode.kt
package nl.pindab0ter.common.helpers

import com.github.kittinunf.fuel.Fuel
import com.github.kittinunf.fuel.core.awaitUnit
import kotlinx.coroutines.runBlocking
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths


/**
 * @param year The year of the puzzle.
 * @param day The day of the puzzle.
 *
 * @return The input for the given year and day.
 */
fun getInput(year: Int, day: Int): String {
    val paddedDay = day.toString().padStart(2, '0')
    val file = Paths.get("src/main/resources/${year}/day$paddedDay/input").toFile()

    if (!file.exists()) runBlocking {
        if (!file.parentFile.exists()) Files.createDirectories(file.parentFile.toPath())

        val sessionCookie = File(".session-cookie").readText()

        Fuel
            .download("https://adventofcode.com/${year}/day/${day}/input")
            .fileDestination { response, _ ->
                if (response.statusCode == 400) throw Exception("Invalid session cookie")
                if (response.statusCode != 200) throw Exception("Unexpected status code: ${response.statusCode}, ${response.responseMessage}")

                file
            }
            // https://www.reddit.com/r/adventofcode/comments/z9dhtd/please_include_your_contact_info_in_the_useragent/
            .appendHeader(
                "User-Agent" to "https://github.com/pindab0ter/AdventOfCode",
                "From" to "hansvanluttikhuizen@me.com",
                "Cookie" to "session=$sessionCookie",
            )
            .awaitUnit()
    }

    return file.readText().trimEnd()
}

File: src/main/clojure/nl/pindab0ter/common/advent_of_code.clj
(ns nl.pindab0ter.common.advent-of-code
  (:require
    [clj-http.client :as http]
    [clojure.java.io :as io]
    [clojure.string :as str]))


(defn get-input
  "Returns the input for the given year and day."
  [year day]
  (let [file (io/file (str "src/main/resources/" year "/day" (format "%02d" day) "/input"))]
    (io/make-parents file)
    (if-not (.exists file)
      (let [session-cookie (slurp ".session-cookie")
            ;; https://www.reddit.com/r/adventofcode/comments/z9dhtd/please_include_your_contact_info_in_the_useragent/
            options        {:headers              {"User-Agent" "https://github.com/pindab0ter/AdventOfCode"
                                                   "From"       "hansvanluttikhuizen@me.com"
                                                   "Cookie"     (str "session=" session-cookie)}
                            :as                   :byte-array
                            :unexceptional-status #(or (<= 200 % 299) (= 400 %))}
            response       (http/get (str "https://adventofcode.com/" year "/day/" day "/input") options)
            status         (:status response)
            body           (str/trim-newline (slurp (:body response)))]
        (cond
          (= 400 status) (throw (Exception. "Invalid session cookie. Please provide a valid session cookie in .session-cookie."))
          (not= 200 status) (throw (Exception. (str "Unexpected status code: " (:status response) "\n" (slurp (:body response)))))
          :else (spit file body))))
    (slurp file)))
